karapace/kafka_rest_apis/__init__.py

```diff
@@ -477,27 +477,24 @@ async def _maybe_create_async_producer(self) -> AIOKafkaProducer:
 
                 log.info("Creating async producer")
 
-                # Don't retry if creating the SSL context fails, likely a configuration issue with
-                # ciphers or certificate chains
-                ssl_context = create_client_ssl_context(self.config)
-
-                # Don't retry if instantiating the producer fails, likely a configuration error.
-                producer = AIOKafkaProducer(
-                    acks=acks,
-                    bootstrap_servers=self.config["bootstrap_uri"],
-                    compression_type=self.config["producer_compression_type"],
-                    connections_max_idle_ms=self.config["connections_max_idle_ms"],
-                    linger_ms=self.config["producer_linger_ms"],
-                    max_request_size=self.config["producer_max_request_size"],
-                    metadata_max_age_ms=self.config["metadata_max_age_ms"],
-                    security_protocol=self.config["security_protocol"],
-                    ssl_context=ssl_context,
-                    **get_kafka_client_auth_parameters_from_config(self.config),
-                )
-
                 try:
+                    producer = AsyncKafkaProducer(
```

Comment:
suggestion: It looks like this try-block aught to have a `finally: producer.stop()`?

Reply:
Now, this makes me wonder whether this is the cause of these too many open files errors we have in sentry:

https://aiven-io.sentry.io/issues/4625989357/?environment=production&project=1219632&query=is%3Aunresolved+assigned%3A%23team-helpful-husky&referrer=issue-stream&statsPeriod=30d&stream_index=2

Reply:
> suggestion: It looks like this try-block aught to have a finally: producer.stop()?
I think if we do that every time we start the producer we would also stop it right away. We could call stop in the except clause, but if the underlying `KafkaProducer` cannot be created the thread will not be started, so calling stop would do nothing.

If I read that Sentry error correctly it's around admin client creation in Karapace version 3.9 (although maybe it's present for other versions but Sentry is not very helpful there) 🤔

Reply:
Ah yes, you're right, it's only the exception cases we should close it.

> but if the underlying KafkaProducer cannot be created the thread will not be started, so calling stop would do nothing

I don't think we can say much about which resources were successfully created if an exception occurs here. For the caught ones, it definitely looks like we can have opened a socket that would be abandoned and lingering if not cleaned up, but same goes for any exception not explicitly handled.